### ISSUE TITLE
Cannot write dataframe to parquet in S3 bucket

### ISSUE THREAD

**Issue by mdwint** [dask/s3fs#352]
When trying to write a Dask dataframe to S3 using `to_parquet`, s3fs tries to create the bucket even though it exists. Reading from the bucket does work.

The issue occurs on s3fs 0.5.0, but not on 0.4.2.

**Minimal Complete Verifiable Example**:

```python
>>> import dask.dataframe as dd
>>> df = dd.read_parquet('s3://my-bucket/input/')
>>> df
Dask DataFrame Structure:
...
Dask Name: read-parquet, 1 tasks

>>> df.to_parquet('s3://my-bucket/output/')
Traceback (most recent call last):
  File "/Users/matteo/.local/share/virtualenvs/gorilla-engine-kOUz7rMz/lib/python3.7/site-packages/s3fs/core.py", line 446, in _mkdir
    await self.s3.create_bucket(**params)
  File "/Users/matteo/.local/share/virtualenvs/gorilla-engine-kOUz7rMz/lib/python3.7/site-packages/aiobotocore/client.py", line 134, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (IllegalLocationConstraintException) when calling the CreateBucket operation: The unspecified location constraint is incompatible for the region specific endpoint this request was sent to.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/matteo/.local/share/virtualenvs/gorilla-engine-kOUz7rMz/lib/python3.7/site-packages/dask/dataframe/core.py", line 3950, in to_parquet
    return to_parquet(self, path, *args, **kwargs)
  File "/Users/matteo/.local/share/virtualenvs/gorilla-engine-kOUz7rMz/lib/python3.7/site-packages/dask/dataframe/io/parquet/core.py", line 473, in to_parquet
    **kwargs_pass,
  File "/Users/matteo/.local/share/virtualenvs/gorilla-engine-kOUz7rMz/lib/python3.7/site-packages/dask/dataframe/io/parquet/arrow.py", line 879, in initialize_write
    fs.mkdirs(path, exist_ok=True)
  File "/Users/matteo/.local/share/virtualenvs/gorilla-engine-kOUz7rMz/lib/python3.7/site-packages/fsspec/spec.py", line 1018, in mkdirs
    return self.makedirs(path, exist_ok=exist_ok)
  File "/Users/matteo/.local/share/virtualenvs/gorilla-engine-kOUz7rMz/lib/python3.7/site-packages/s3fs/core.py", line 460, in makedirs
    self.mkdir(path, create_parents=True)
  File "/Users/matteo/.local/share/virtualenvs/gorilla-engine-kOUz7rMz/lib/python3.7/site-packages/fsspec/asyn.py", line 100, in wrapper
    return maybe_sync(func, self, *args, **kwargs)
  File "/Users/matteo/.local/share/virtualenvs/gorilla-engine-kOUz7rMz/lib/python3.7/site-packages/fsspec/asyn.py", line 80, in maybe_sync
    return sync(loop, func, *args, **kwargs)
  File "/Users/matteo/.local/share/virtualenvs/gorilla-engine-kOUz7rMz/lib/python3.7/site-packages/fsspec/asyn.py", line 51, in sync
    raise exc.with_traceback(tb)
  File "/Users/matteo/.local/share/virtualenvs/gorilla-engine-kOUz7rMz/lib/python3.7/site-packages/fsspec/asyn.py", line 35, in f
    result[0] = await future
  File "/Users/matteo/.local/share/virtualenvs/gorilla-engine-kOUz7rMz/lib/python3.7/site-packages/s3fs/core.py", line 450, in _mkdir
    raise translate_boto_error(e) from e
PermissionError: The unspecified location constraint is incompatible for the region specific endpoint this request was sent to.
```

**Environment**:

- Dask version: 2.24.0
- s3fs version: 0.5.0
- botocore version: 1.17.44
- Python version: 3.7.4
- Operating System: macOS 10.15.6
- Install method (conda, pip, source): pip

**Comment by martindurant** [comment]
Hm, it looks like you are trying to create a bucket, but the bucket already exists.
I seem to be perfectly able to call mkdir when the bucket already exists, so I can only suggest that you default region doesn't match the region that the bucket has (parameter `region_name` in `client_kwargs`).

So there are two things here:
- we can check for the existence of the bucket, and don't try to create if not necessary
- we can allow mkdir to fail
(note that AWS permissioning is complex, so it's possible that you have the ability to create buckets, but not list a particular bucket, so the only way to know that it exists is to try to create it and fail).

**Comment by mdwint** [comment]
Stepping through this in a debugger, I noticed that `_mkdir` on 0.5.0 was different from master. It looks like this may have been fixed in #347.

**Comment by mdwint** [comment]
I'll wait for the next release! Sorry for the confusion.
